### DISCUSSION TITLE
Why do setters use `Args extends unknown[]` and `...args: Args` for their types when you only pass one value?

### DISCUSSION THREAD

**Discussion by AriaFallah** [pmndrs/jotai#2523] [Q&A]
I'm really confused because the types are currently written as

https://github.com/pmndrs/jotai/blob/7a68a1246f1c174af3dd2979de202aa6869db7a7/src/vanilla/atom.ts#L3-L10

But why is the value after atom `...args: Args`? I cannot find a single example of `set` called with more than one argument. Why is it not instead something like `newValue: Value | ((oldValue: Value) => Value)`?

**Comment by dai-shi** [comment]
Because we support multiple arguments. This is a huge improvement in Jotai v2.
Yeah, we may not have any examples in docs. Would you mind adding one?

For primitive atoms we set the type like this:
https://github.com/pmndrs/jotai/blob/7a68a1246f1c174af3dd2979de202aa6869db7a7/src/vanilla/atom.ts#L59-L65

**Reply by AriaFallah** [reply]
I would, but I am new to `jotai` so I have no idea what an example would even look like 😃 . I was asking because I was running into something similar to this https://github.com/pmndrs/jotai/issues/1452, but not exactly. I can try and figure out a repro so you can see

**Reply by AriaFallah** [reply]
Here's the [tsplay link](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgQxhEAaOBxApjGXKLAZX0KjgF84AzKdOAIgCsIZlgmBuAKF4DGEAHYBneOwAWRAIJoQcALwp5AClrIANqNxZVAc3xYdMAJRKAfIl5w4J1VNnysMKAFdcp3lS+9HUOXQAOhEAWQg3YXhlVRNAkHNFKwQbO3x41S8qXiA). Not sure if what I'm doing is illegal, which is why it doesn't typecheck. [Here's another example](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgQxhEAaOBxApjGXKLAZX0KjgF84AzKdOAIgCsIZlgmBuAKF4DGEAHYBneOwAWRAIJoQcALwp5AClrIANqNxZVAc3xYdMAJRKAfIiqnBI8XELi5jZanQH8luIZiqpsvKmer7G+OaKVgi8cHAm-jDSUC6YjlAArri2NrxAA). I guess you cannot call `set` on an atom that specifies a write function? Unless [you annotate it properly that is](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgQxhEAaOBxApjGXKLAZX0KjgF84AzKdOAIgCsIZlgmBuAKF4DGEAHYBneOwAWRAIJoQcALwp5AHgBGECABtcyYVgDamnXuEBdLADcIwACYA+ABS1k20bixOA5viweYAEolB0QqQMERcThCcTlGZVR0H3wQuF8YJylZeUCvDP98YMVQhF44OACsmGkoeMwYqABXXAjw3iA).

**Reply by dai-shi** [reply]
https://tsplay.dev/weAnEw this also works.

The first one is kind of illegal. TS can't type it. We sometimes use it for a hack, but it's not a public API usage.

**Reply by AriaFallah** [reply]
Btw you wrote

> Because we support multiple arguments. This is a huge improvement in Jotai v2.

What's the improvement and usecase? I'm having difficulty imagining it to be honest. If I ever wanted multiple arguments, I would either pass an array or an object. How's varargs offer an advantage over this?

**Reply by dai-shi** [reply]
Well, it's just not for you. People had seemed it's a huge limitation in v1. That's why I said a huge improvement, but it's not something widely used. I only used it once or twice. Don't remember where. Passing an array works and I suggested so in v1.

**Reply by dai-shi** [reply]
This also supports zero arguments, which wasn't possible in v1. Types are cleaner in this sense too, compared to v1.

**Reply by AriaFallah** [reply]
> Well, it's just not for you

Haha of course. Hope I'm not coming off that way. I'm mostly just curious about the design decisions and what informed them as I learn the library. Varargs APIs make things way way more complex, and wanted to understand the tradeoff that led to that.

**Reply by dai-shi** [reply]
There's no drawback with `...args` except for the types are harder to read. (but it also avoids some hacks, so actually it makes it easier to read for some cases.)
